tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,219 @@
+     * However, since this buffer is in reality a 'struct cmsghdr', use a
+     * union to ensure that it is aligned as required for that structure.
+     */
+    union {
+        char buf[CMSG_SPACE(sizeof(record_type))];
+        /* Space large enough to hold a ucred structure */
+        struct cmsghdr align;
+    } control_msg;
+
+    msg.msg_control = control_msg.buf;
+    msg.msg_controllen = sizeof(control_msg.buf);
+#endif
+
+    RESULT_GUARD(s2n_ktls_send_control_msg(sock, &msg, record_type, blocked, result));
+
+    RESULT_GUARD(s2n_ktls_send_msg_impl(sock, &msg, msg_iov, blocked, result));
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_recv_msg_impl(int sock, struct msghdr *msg,
+        struct iovec *msg_iov, s2n_blocked_status *blocked, ssize_t *result)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(result);
+
+    /* set receive buffer */
+    msg->msg_iov = msg_iov;
+    msg->msg_iovlen = 1;
+
+    *blocked = S2N_BLOCKED_ON_READ;
+    *result = recvmsg(sock, msg, 0);
+    if (*result == -1) {
+        if (errno == EWOULDBLOCK || errno == EAGAIN) {
+            RESULT_BAIL(S2N_ERR_IO_BLOCKED);
+        }
+        RESULT_BAIL(S2N_ERR_IO);
+    }
+    *blocked = S2N_NOT_BLOCKED;
+
+    /* The return value will be 0 when the peer has performed an orderly shutdown. */
+    RESULT_ENSURE(*result != 0, S2N_ERR_CLOSED);
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_recv_control_msg(int sock, struct msghdr *msg,
+        uint8_t *record_type, s2n_blocked_status *blocked, ssize_t *result)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(record_type);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(result);
+
+#if S2N_KTLS_SUPPORTED
+    /* attempt to read the ancillary data */
+    struct cmsghdr *hdr = CMSG_FIRSTHDR(msg);
+    RESULT_ENSURE(hdr != NULL, S2N_ERR_IO);
+
+    if (hdr->cmsg_level == S2N_SOL_TLS && hdr->cmsg_type == S2N_TLS_GET_RECORD_TYPE) {
+        *record_type = *(unsigned char *) CMSG_DATA(hdr);
+    } else {
+        /* if no RECORD_TYPE is included then assume its TLS_APPLICATION_DATA.
+         * This is ok since the socket has kTLS enabled and all data arriving
+         * must have been encrypted with the correct keys.
+         */
+        *record_type = TLS_APPLICATION_DATA;
+    }
+#endif
+
+    return S2N_RESULT_OK;
+}
+
+/* Best practices taken from
+ * https://man7.org/tlpi/code/online/dist/sockets/scm_cred_recv.c.html */
+S2N_RESULT s2n_ktls_recv_msg(int sock, uint8_t *buf, size_t length,
+        uint8_t *record_type, s2n_blocked_status *blocked,
+        ssize_t *result)
+{
+    RESULT_ENSURE_REF(buf);
+    RESULT_ENSURE_REF(record_type);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(result);
+    RESULT_ENSURE_GT(length, 0);
+
+    /* Init msghdr */
+    struct msghdr msg = { 0 };
+    msg.msg_name = NULL;
+    msg.msg_namelen = 0;
+
+#if S2N_KTLS_SUPPORTED
```

Comment:
Does this actually need to be gated on this?

Reply:
The `CMSG_SPACE` macro is platform specific and I was seeing errors on other platforms